src/workerd/io/actor-cache.c++

```diff
@@ -2139,6 +2142,37 @@ void ActorCache::ensureFlushScheduled(const WriteOptions& options) {
   }
 }
 
+kj::Maybe<kj::Promise<void>> ActorCache::onNoPendingFlush() {
+  // This function returns a Maybe<Promise> because a falsy maybe allows the jsg interface to make
+  // a resolved jsg::Promise. This is meaningfully different from a ready kj::Promise because it
+  // allows the next continuation to run immediately on the microtask queue instead of returning to
+  // the kj event loop and fulfilling a resolver that enqueues the continuation.
+
+  if (lru.options.neverFlush) {
+    // We won't ever flush (usually because we're a preview session), so return a falsy maybe.
+    return nullptr;
+  }
+
+  if (flushScheduled) {
+    // Our current run has done operations that need to be flushed, we need to wait for that flush
+    // to complete before resolving the jsg::Promise.
+    return lastFlush.addBranch();
+  }
+
+  if (flushesEnqueued > 0) {
+    // Some previous run had done operations that needed to be flushed and it is still ongoing. This
+    // implies that `allowConcurrency` was set to true for those operations since the input lock did
```

Comment:
This isn't really true, is it? Puts having been buffered for flushing does not take the input lock. `allowConcurrency` only affects writes if backpressure kicks in, and even there the effect is pretty minimal.

Reply:
Hmmmm, it's actually more about outstanding input gate-required operations being sequenced with the writes. I'll rewrite the comment to explicitly call out that the input gate is the important part.